content/blog/2022-05-10-compcert/index.md

```diff
@@ -0,0 +1,112 @@
+  volleyball and cooking.
+"""
+[[extra.authors]]
+name = "Anshuman Mohan"
+link = "https://www.cs.cornell.edu/~amohan/"
+[[extra.authors]]
+name = "Alaia Solko-Breslin"
+link = "https://alaiasolkobreslin.github.io/"
++++
+
+### Background: Formal Verification
+
+A key reason for the success of the CompCert project is that it foresaw – and triggered – an increasing interest in the formal verification of software. Proponents of formal verification argue that traditional testing is often unsafe and untrustworthy, and that we need more mathematically sound assurances of what our code does. This is generally agreed to be a good idea; the strongest detractors of verification mostly argue that it is too expensive (in time, in maintainability, and therefore in money) to be put into use.
+
+Some of this criticism is totally fair, and no one project can change the landscape. However, CompCert lowers the barrier to entry significantly. Programmers can state and prove claims about their code as written, in the source language. This is far from easy; it is a whole domain of research in CS. However, CompCert then guarantees that these hard-won claims still hold true even after the source code is compiled to an executable. The effect is multiplicative and powerful.
+
+
+### Trusting Your Compiler
+
+In a sense a compiler is very easy to specify: it should transform code without changing its behavior. However, as the author shows, it is critical that we take a more flexible stance to allow for undefined behavior, nondeterminism, and permissible changes to the behavior of the source code. Indeed, the wise programmer is aiming not for identical behavior, but rather for behavior that is close enough to the original, for some choice of “close enough”. 
+
+This is intentionally open-ended. Loosely speaking, choosing a strong definition of “close enough” leads to more work and stronger guarantees, while a more relaxed choice leads to less work and weaker guarantees. In class we discussed that this vagueness, in addition to the vagueness about “going wrong”, is a feature of this work. 
+
+The author provides five ways to think about a compiler from a semantic perspective:
+
+1. $ \forall B, S \Downarrow B \Leftrightarrow C \Downarrow B $
+2. $ S \text{ safe} \Rightarrow (\forall B, C \Downarrow B \Rightarrow S \Downarrow B) $
+3. $ \forall B \notin \text{Wrong}, S \Downarrow B \Rightarrow C \Downarrow B $ 
+4. $ S \vDash \text{Spec} \Rightarrow C \vDash \text{Spec} $
+5. $ S \text{ safe} \Rightarrow C \text{ safe} $ 
+
+
+These are not interchangeable, but the author explains how they relate to each other and points out features (such as transitivity) that they have in common. In class we took a deeper look at $ S \vDash \text{Spec} \Rightarrow C \vDash \text{Spec} $. While CompCert does not explicitly use that flavor of semantic preservation in its proof, this is arguably the easiest way to appreciate CompCert’s overall goal.
```

Comment:
Writing down an intuitive/informal (English) translation of this math seems like it would be useful. Maybe doing this for all 5 options would be too much, but doing it for the one(s) CompCert actually embraces could be nice. (As would explaining its advantages/disadvantages over the others.)